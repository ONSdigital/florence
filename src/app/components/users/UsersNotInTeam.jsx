import React, { useState } from "react";
import { connect } from "react-redux";
import DynamicList from "../../components/dynamic-list/DynamicList";
import PropTypes from "prop-types";
import { selectNewTeamAllPreviewUsers } from "../../config/selectors";

const propTypes = {
    loading: PropTypes.bool,
    addUserToTeam: PropTypes.func,
    usersNotInTeam: PropTypes.arrayOf(PropTypes.object),
    allPreviewUsers: PropTypes.arrayOf(PropTypes.object),
};
const UsersNotInTeam = props => {
    const { allPreviewUsers, usersNotInTeam, addUserToTeam } = props;
    const [filterTerm, setFilterTerm] = useState("");
    const handleSearch = event => {
        const searchTerm = event.target?.value?.toLowerCase();
        setFilterTerm(searchTerm);
    };
    const compare = (a, b) => {
        return a.title?.localeCompare(b.title);
    };

    let usersNotInTeamList = [];
    if (usersNotInTeam) {
        usersNotInTeamList = usersNotInTeam
            .map(user => {
                return {
                    title: `${user.forename} ${user.lastname}`,
                    desc: user.email,
                    icon: "Person",
                    buttonName: "Add",
                    buttonCallback: () => {
                        const newUser = allPreviewUsers.find(viewer => viewer.email === user.email);
                        addUserToTeam(newUser);
                    },
                    iconColor: "standard",
                };
            })
            .filter(viewer => viewer.title?.toLowerCase().search(filterTerm) !== -1 || viewer.desc?.toLowerCase().search(filterTerm) !== -1)
            .sort(compare);
    }
    return (
        <div className="grid__col-4 margin-top--3">
            <DynamicList
                title="Add member to team"
                headingLevel="2"
                listItems={usersNotInTeamList}
                noResultsText="No viewers found"
                filterPlaceholder="Search viewers by name or email"
                handleSearchInput={handleSearch}
                listHeightClass="height--60"
                loading={props.loading}
            />
        </div>
    );
};

UsersNotInTeam.propTypes = propTypes;

function mapStateToProps(state) {
    return {
        allPreviewUsers: selectNewTeamAllPreviewUsers(state),
    };
}

export default connect(mapStateToProps)(UsersNotInTeam);
