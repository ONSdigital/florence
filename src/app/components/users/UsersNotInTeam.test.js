import React from "react";
import UsersNotInTeam from "./UsersNotInTeam";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import { WrapperComponent } from "../../../tests/test-utils";

let dispatchedActions = [];
const mockDispatch = event => {
    dispatchedActions.push(event);
};

describe("UsersNotInTeam", () => {
    const UsersNotInTeamProps = {
        dispatch: mockDispatch,
        loading: false,
    };
    describe("Given valid props on initial load", () => {
        const component = mount(
            <WrapperComponent>
                <UsersNotInTeam {...UsersNotInTeamProps} />
            </WrapperComponent>
        );
        it("renders the relevant components to screen", () => {
            expect(component.find("h2.dynamic-list__title")).toHaveLength(1);
            expect(component.find("input#add-member-to-team-search-content-types")).toHaveLength(1);
        });
        it("matches snapshot", () => {
            const componentForSnapshot = renderer.create(
                <WrapperComponent>
                    <UsersNotInTeam {...UsersNotInTeamProps} />
                </WrapperComponent>
            );
            expect(componentForSnapshot.toJSON()).toMatchSnapshot();
        });
    });
});
