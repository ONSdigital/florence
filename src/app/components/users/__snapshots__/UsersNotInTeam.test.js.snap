// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersNotInTeam Given valid props on initial load matches snapshot 1`] = `
<div
  className="grid__col-4 margin-top--3"
>
  <div>
    <h2
      className="dynamic-list__title"
    >
      Add member to team
    </h2>
    <label
      className="visually-hidden"
      htmlFor="add-member-to-team-search-content-types"
    >
      Search viewers by name or email
    </label>
    <div
      className="form__input"
    >
      <label
        className="form__label"
        htmlFor="add-member-to-team-search-content-types"
      />
      
      <input
        autoComplete="new-password"
        autoFocus={false}
        className="input"
        disabled={false}
        id="add-member-to-team-search-content-types"
        name="add-member-to-team-search-content-types"
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search viewers by name or email"
        type="text"
      />
      
    </div>
    <div
      aria-live="assertive"
    >
      No viewers found
    </div>
  </div>
</div>
`;
